Bump python minimum version to 3.8

maflcko opened the pull request:
There is no pressing reason to drop support for 3.7, however there are several maintenance issues:

* There is no supported operating system that ships 3.7 by default. (debian:buster is EOL and unmaintained to the extent that it doesn't run in the CI environment. See https://github.com/bitcoin/bitcoin/pull/27340#issuecomment-1484988445)
* Compiling python 3.7 from source is also unsupported on at least macos, according to https://github.com/bitcoin/bitcoin/pull/24017#issuecomment-1107820790
* Recent versions of lief require 3.8, see https://github.com/bitcoin/bitcoin/pull/27507#issuecomment-1517561645

Fix all maintenance issues by bumping the minimum.

stickies-v commented on doc/dependencies.md:
```diff
@@ -10,7 +10,7 @@ You can find installation instructions in the `build-*.md` file for your platfor
 | [Clang](https://clang.llvm.org) | [8.0](https://github.com/bitcoin/bitcoin/pull/24164) |
 | [GCC](https://gcc.gnu.org) | [8.1](https://github.com/bitcoin/bitcoin/pull/23060) |
-| [Python](https://www.python.org) (tests) | [3.7](https://github.com/bitcoin/bitcoin/pull/26226) |
+| [Python](https://www.python.org) (scripts, tests) | [3.8](https://github.com/bitcoin/bitcoin/pull/xyz) |
```
```suggestion
| [Python](https://www.python.org) (scripts, tests) | [3.8](https://github.com/bitcoin/bitcoin/pull/27483) |
```

stickies-v reviewed (commented):
Concept ACK 

I'm not sure why we're switching to focal instead of bullseye for the `qt5` ci, but I don't really have a view either way. Just pointing it out.

Commit:
test: Use python3.8 pow()

Commit:
Bump python minimum version to 3.8

Also, switch ci_native_qt5 to g++-9 (from g++-8) to work around bugs.
This should be fine, because the i686_centos task still checks for g++-8
compatibility.

See
https://github.com/bitcoin/bitcoin/pull/27483#issuecomment-1513477050
for the list of bugs.

fanquake commented:
Concept ACK

jamesob commented:
Concept ACK

jamesob commented:
It looks like the commit message on https://github.com/bitcoin/bitcoin/pull/27483/commits/fa762d7f52db9c65df967f9cb22c6be963ce0300 is out of date, which suggests we're swiching from g++-8 to clang, but the contents of the commit suggest we're going to g++-9 instead.

sedited reviewed (approved):
utACK fa6eb6516727a8675dc6e46634d8343e282528ab

fanquake commented:
Should also bump the lint DOckerfile to `python:3.8-buster`?

fanquake commented:
> buster is EOL and unmaintained,

In this context, I don't think that makes any difference, as the container is just an interpreter, so as long as it's downloadable, it should work (no installing packages etc).

> I don't really understand why the lint Dockerfile isn't simply using the exact same distro and setup like the lint CI.

No idea. Maybe ask @jamesob.

I was mostly pointing it out because it seemed odd to not just bump it as well. Not a blocker if someone else is going to followup.

Commit:
ci: Bump ci/lint/Dockerfile

This bump should not be needed, see discussion starting at
https://github.com/bitcoin/bitcoin/pull/27483#issuecomment-1517739626

fanquake reviewed (approved):
ACK fac395e5eb2cd3210ba6345f777a586a9bec84e3

fanquake commented:
> It calls RUN /install.sh, which installs packages

Right. Although it doesn't actually need to do that, as all the packages we needs are already available in the base python3.8-buster image.

RandyMcMillan commented:
https://github.com/bitcoin/bitcoin/pull/27130

fjahr commented:
ACK fac395e5eb2cd3210ba6345f777a586a9bec84e3

This is ok to merge but when I wrote that comment I meant to remove the `modinv()` function completely, e.g. https://github.com/fjahr/bitcoin/commit/75b8ba524d3dc957910bc8c0a4d1dd2b8ceaa426. You can pull this commit in here or we merge this now and I will open it as a follow-up. I think the comment in `modinv()` isn't accurate anymore after this change because python seems to use Exponentiation by Squaring and we don't really need a one-line function that and we also don't need to test native python code.

fanquake commented:
> I think it is best for you to open a follow-up, 

Lets do that.
